### PR TITLE
Feature/282 add functionality that save and update information about sitecore environments

### PR DESCRIPTION
**Pull request by DmitryKolinchuk**
This pull request also contains changes related to the following feature:
[https://github.com/Sitecore/Sitecore-Instance-Manager/issues/247](https://github.com/Sitecore/Sitecore-Instance-Manager/issues/247)

### CODE REVIEW

File: src/SIM.Instances/Instance.cs [hunk 1]
```diff
@@ -217,6 +218,16 @@ public virtual string PackagesFolderPath
       }
     }
 
+
+    [NotNull]
+    public virtual SitecoreEnvironment SitecoreEnvironment
+    {
+      get
+      {
+        return new SitecoreEnvironment(this.Name, SitecoreEnvironmentHelper.SitecoreEnvironments);
```
Comment (IvanSheyenko) [inline src/SIM.Instances/Instance.cs]: every call to the property creates a new object carrying the same data.

File: src/SIM.Instances/InstanceManager.cs [hunk 2]
```diff
@@ -113,6 +114,8 @@ public Instance GetInstance([NotNull] string name)
 
     public void Initialize([CanBeNull] string defaultRootFolder = null)
     {
+      SitecoreEnvironmentHelper.SitecoreEnvironments = SitecoreEnvironmentHelper.GetSitecoreEnvironmentData();
```
Comment (IvanSheyenko) [inline src/SIM.Instances/InstanceManager.cs]: SitecoreEnvironments should be initialized inside SitecoreEnvironmentHelper class

File: src/SIM.Pipelines/Install/GenerateSitecoreEnvironmentData.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        InstallParam installParam = powerShellTask.LocalParams.FirstOrDefault(x => x.Name == SiteName);
+        if (installParam != null && !string.IsNullOrEmpty(installParam.Value))
+        {
+          Hashtable evaluatedLocalParams = tasker.EvaluateLocalParams(powerShellTask.LocalParams, tasker.GlobalParams);
+          if (evaluatedLocalParams != null && evaluatedLocalParams[SiteName] != null)
+          {
+            SitecoreEnvironmentMember sitecoreEnvironmentMember =
+              new SitecoreEnvironmentMember(evaluatedLocalParams[SiteName].ToString(),
+                SitecoreEnvironmentMember.Types.Site.ToString());
+            if (!sitecoreEnvironment.Members.Contains(sitecoreEnvironmentMember))
```
Comment (IvanSheyenko) [inline src/SIM.Pipelines/Install/GenerateSitecoreEnvironmentData.cs]: Contains compares objects via equals method, which by default is done comparing hashcodes, so members with the same name and id but different objects are not equal.

File: src/SIM.SitecoreEnvironments/SitecoreEnvironmentHelper.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using Newtonsoft.Json;
+
+namespace SIM.SitecoreEnvironments
+{
+  public static class SitecoreEnvironmentHelper
```
Comment (IvanSheyenko) [inline src/SIM.SitecoreEnvironments/SitecoreEnvironmentHelper.cs]: The class should encapsulate list of environments all manipulations with the list should be done via this class. It should not be possible to add an environment/member without validation.

File: src/SIM.SitecoreEnvironments/SitecoreEnvironmentHelper.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    private const string FileName = "Environments.json";
+
+    public static string FilePath
+    {
+      get { return Path.Combine(ApplicationManager.ProfilesFolder, FileName); }
+    }
+
+    public static List<SitecoreEnvironment> SitecoreEnvironments;
+
+    public static List<SitecoreEnvironment> GetSitecoreEnvironmentData()
```
Comment (IvanSheyenko) [inline src/SIM.SitecoreEnvironments/SitecoreEnvironmentHelper.cs]: We have 'public static List<SitecoreEnvironment> SitecoreEnvironments' but GetSitecoreEnvironmentData always reads from file.

**Comment by DmitryKolinchuk** [comment]
@IvanSheyenko I updated the code according to your suggestions. Please check it.

File: src/SIM.SitecoreEnvironments/SitecoreEnvironment.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  {
+    public SitecoreEnvironment() { }
+
+    public SitecoreEnvironment(string instanceName)
+    {
+      ID = Guid.NewGuid();
+      Name = instanceName;
+    }
+
+    public SitecoreEnvironment(string instanceName, List<SitecoreEnvironment> sitecoreEnvironments)
```
[the remark this answers is not in the document]
Comment (DmitryKolinchuk) [reply src/SIM.SitecoreEnvironments/SitecoreEnvironment.cs]: @IvanSheyenko I refactored the code. Please review the changes.

**Pull request merged**